debezium-core/src/main/java/io/debezium/pipeline/ChangeEventSourceCoordinator.java

```diff
@@ -122,8 +126,23 @@ public ChangeEventSourceCoordinator(OffsetContext previousOffset, ErrorHandler e
         });
     }
 
+    protected CatchUpStreamingResult executeCatchUpStreaming(OffsetContext previousOffset, ChangeEventSourceContext context,
+                                                             SnapshotChangeEventSource snapshotSource)
+            throws InterruptedException {
+        return new CatchUpStreamingResult(false);
+    }
+
+    protected void streamEvents(OffsetContext offsetContext, ChangeEventSourceContext context) throws InterruptedException {
+        streamingSource = changeEventSourceFactory.getStreamingChangeEventSource(offsetContext);
+        eventDispatcher.setEventListener(streamingMetrics);
+        streamingMetrics.connected(true);
+        LOGGER.info("Starting streaming");
+        streamingSource.execute(context);
+        LOGGER.info("Finished streaming");
+    }
+
     public void commitOffset(Map<String, ?> offset) {
-        if (streamingSource != null && offset != null) {
+        if (!commitOffsetLock.isLocked() && streamingSource != null && offset != null) {
```

Comment:
@gunnarmorling I’ve been thinking about your comments about synchronization on https://github.com/debezium/debezium/pull/1596 and I think I still need some clarification.  Just wrapping all access to the `streamingSource` with synchronization blocks inline in the `ChangeEventSourceCoordinator#start` and `ChangeEventSourceCoordinator#commitOffset` methods wouldn’t protect from a race condition between the offset commit and the catch up streaming deactivation by setting `streamingSource` to `null`.  

There are a couple options that I’ve looked at other than using the `ReentrantLock`:

A new synchronized function could be added here that handles both the null check and `streamingSource.commitOffset(offset)` call in `ChangeEventSourceCoordinator#commitOffset` and the `streamingSource = null` deactivation in `ChangeEventSourceCoordinator#start`. Personally, this feels unnecessarily complex and would be difficult to integrate with the other calls to `streamingSource`.

OR

A separate object stored in an instance variable could be use as a lock to synchronize on.
```
Object streamingSourceLock = new Object();
 ...
public void commitOffset(Map<String, ?> offset) {
    synchronized (streamingSourceLock) {
        if (streamingSource != null && offset != null) {
            streamingSource.commitOffset(offset);
        }
    }
}
```
Since `BaseSourceTask` has access to the executing `ChangeEventSourceCoordinator` reference, both threads should synchronize correctly. That said, I feel using a generic `Object` as a lock is unidiomatic compared to the `ReentrantLock` and I like the `ReentrantLock` because it highlights where the race condition might occur.

Thoughts?

Reply:
Getting back to this, now that 1.2 Final is out :)

Good point, if the object can be set to null, we cannot synchronize on it. An alternative would be to declare the methods themselves as sychronized (or extract reasonably scoped methods), in which case the entire coordinator instance itself would be the synchronization object.

Reply:
Could you clarify what you mean by synchronizing the methods? The `start` method is already synchronized and since streaming occurs there, any thread that calls `start` won't leave the connector until the connector shuts down.  Adding another synchronized method to this class sounds like a that would cause a deadlock.

Reply:
@grantcooksey, I was a bit misled and missed that all this code runs async via the separate executor, i.e. the sychronized definition of `start()` doesn't impact it at all. I think your proposed solution is good.